### PR TITLE
[NTOSKRNL] Implement PeakCommitment (MmPeakCommitment, MmTotalCommittedPages)

### PR DESCRIPTION
**Pull request by KRosUser**
## Purpose

PeakComitment is currently not implemented, forced to 0
image

Proposal to implement it 
image

### CODE REVIEW

**Comment by HBelusca** [comment]
Can you show Windows tasmkgr's switch-to in the very same kind of setup (more than one task selected)?

**Comment by KRosUser** [comment]
> Can you show Windows tasmkgr's switch-to in the very same kind of setup (more than one task selected)?

Yes, it follow the principle of the proposal :
image

**Comment by KRosUser** [comment]
UI aspects carved out to [PR](https://github.com/reactos/reactos/pull/4652)

File: ntoskrnl/mm/balance.c [hunk 1]
```diff
@@ -81,6 +93,7 @@ MmReleasePageMemoryConsumer(ULONG Consumer, PFN_NUMBER Page)
     }
 
     (void)InterlockedDecrementUL(&MiMemoryConsumers[Consumer].PagesUsed);
+    UpdatePeakCommitment();
```
[the remark this answers is not in the document]
Comment (KRosUser) [reply ntoskrnl/mm/balance.c:96]: I don't think so. Fixed !

File: ntoskrnl/mm/balance.c [hunk 2]
```diff
@@ -39,6 +39,16 @@ static LONG PageOutThreadActive;
 
 /* FUNCTIONS ****************************************************************/
 
+static inline VOID UpdatePeakCommitment(VOID)
+{
+    // HACK: MmTotalCommittedPages should be adjusted consistently with
+    // other counters at different places.
+    MmTotalCommittedPages = MiMemoryConsumers[MC_SYSTEM].PagesUsed + MiMemoryConsumers[MC_USER].PagesUsed + MiUsedSwapPages;
```
Comment (SergeGautherie) [inline ntoskrnl/mm/balance.c:46]: Why no update call after `MiUsedSwapPages` increase?

**Comment by KRosUser** [comment]
> Can you show what Windows TaskMgr shows at the same time?

Same data :
image

File: ntoskrnl/mm/balance.c (hunk 2 above)
Comment (KRosUser) [reply ntoskrnl/mm/balance.c:46]: Good point. Fixed

**Comment by KRosUser** [comment]
@HBelusca : I see you still have the "change requested" flag. Did I miss something ?

File: ntoskrnl/mm/balance.c [hunk 3]
```diff
@@ -39,6 +39,16 @@ static LONG PageOutThreadActive;
 
 /* FUNCTIONS ****************************************************************/
 
+static inline VOID UpdatePeakCommitment(VOID)
+{
+    // HACK: MmTotalCommittedPages should be adjusted consistently with
+    // other counters at different places.
```
Comment (SergeGautherie) [inline ntoskrnl/mm/balance.c:45]: Is this comment still relevant? If yes, which other counters/places would you have in mind, for example?

Comment (KRosUser) [reply ntoskrnl/mm/balance.c:45]: Was added by @HBelusca . @HBelusca : What's your view on this comment now ?

File: ntoskrnl/mm/balance.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (KRosUser) [reply ntoskrnl/mm/balance.c:46]: Right

File: ntoskrnl/mm/balance.c [hunk 4]
```diff
@@ -50,6 +60,8 @@ MmInitializeBalancer(ULONG NrAvailablePages, ULONG NrSystemPages)
     MiMinimumAvailablePages = 256;
     MiMinimumPagesPerRun = 256;
     MiMemoryConsumers[MC_USER].PagesTarget = NrAvailablePages / 2;
+    MmPeakCommitment = 0;
+    UpdatePeakCommitment();
```
[the remark this answers is not in the document]
Comment (KRosUser) [reply ntoskrnl/mm/balance.c:64]: Because i didn't knew globals were implicitely initialized at 0x0

File: ntoskrnl/mm/balance.c [hunk 5]
```diff
@@ -50,6 +60,8 @@ MmInitializeBalancer(ULONG NrAvailablePages, ULONG NrSystemPages)
     MiMinimumAvailablePages = 256;
     MiMinimumPagesPerRun = 256;
     MiMemoryConsumers[MC_USER].PagesTarget = NrAvailablePages / 2;
+    MmPeakCommitment = 0;
```
[the remark this answers is not in the document]
Comment (KRosUser) [reply ntoskrnl/mm/balance.c:63]: I didn't knew. fixed.

File: ntoskrnl/mm/balance.c (hunk 3 above)
Comment (HBelusca) [reply ntoskrnl/mm/balance.c:45]: My view: Since we are not experts in Mm, and this PR isn't an actual serious attempt at getting all the Mm counters consistently implemented, the purpose of this comment is to tell any future Mm-knowledgeable dev to double-check and possibly fix where MmTotalCommittedPages (and the rest) needs to be updated.

**Comment by HBelusca** [comment]
Asking review to @ThFabba @tkreuzer @zefklop regarding the correctness of where the MmTotal...Pages counter needs to be modified in order to fairly represent the current total number of pages allocated.

File: ntoskrnl/mm/balance.c (hunk 3 above)
Comment (KRosUser) [reply ntoskrnl/mm/balance.c:45]: OK so let's keep this comment anyway

**Comment by KRosUser** [comment]
> Asking review to @ThFabba @tkreuzer @zefklop regarding the correctness of where the MmTotal...Pages counter needs to be modified in order to fairly represent the current total number of pages allocated.

Yes, keeping in mind that this  PR isn't an actual serious attempt at getting all the Mm counters consistently implemented,

File: ntoskrnl/ex/sysinfo.c [hunk 6]
```diff
@@ -718,17 +718,15 @@ QSI_DEF(SystemPerformanceInformation)
      *   Add up all the used "Committed" memory + pagefile.
      *   Not sure this is right. 8^\
      */
-    Spi->CommittedPages = MiMemoryConsumers[MC_SYSTEM].PagesUsed +
-                          MiMemoryConsumers[MC_USER].PagesUsed +
-                          MiUsedSwapPages;
+    Spi->CommittedPages = MmTotalCommittedPages;
```
Comment (SergeGautherie) [inline ntoskrnl/ex/sysinfo.c:721]: Calculation is moved, so should its related comment.

File: ntoskrnl/mm/balance.c [hunk 7]
```diff
@@ -39,6 +39,18 @@ static LONG PageOutThreadActive;
 
 /* FUNCTIONS ****************************************************************/
 
+static inline VOID UpdatePeakCommitment(VOID)
```
Comment (SergeGautherie) [inline ntoskrnl/mm/balance.c:42]:
```suggestion
static inline VOID UpdateTotalCommittedPages(VOID)
```
PeakCommitment is a consequence here.

File: ntoskrnl/mm/balance.c [hunk 8]
```diff
@@ -39,6 +39,18 @@ static LONG PageOutThreadActive;
...
+{
+    // HACK: MmTotalCommittedPages should be adjusted consistently with
+    // other counters at different places.
+    MmTotalCommittedPages = MiMemoryConsumers[MC_SYSTEM].PagesUsed +
+                          MiMemoryConsumers[MC_USER].PagesUsed +
+                          MiUsedSwapPages;
```
Comment (SergeGautherie) [inline ntoskrnl/mm/balance.c:48]:
```suggestion
    MmTotalCommittedPages = MiMemoryConsumers[MC_SYSTEM].PagesUsed +
                            MiMemoryConsumers[MC_USER].PagesUsed +
                            MiUsedSwapPages;
```

File: ntoskrnl/mm/pagefile.c [hunk 9]
```diff
@@ -89,6 +89,16 @@ static BOOLEAN MmSystemPageFileLocated = FALSE;
 
 /* FUNCTIONS *****************************************************************/
 
+static inline VOID UpdatePeakCommitment(VOID)
+{
+    // HACK: MmTotalCommittedPages should be adjusted consistently with
+    // other counters at different places.
+    MmTotalCommittedPages = MiMemoryConsumers[MC_SYSTEM].PagesUsed + MiMemoryConsumers[MC_USER].PagesUsed + MiUsedSwapPages;
```
Comment (SergeGautherie) [inline ntoskrnl/mm/pagefile.c:96]: Since you duplicate the function, keep it in sync!

Comment (ThFabba) [reply ntoskrnl/mm/pagefile.c:96]: Don't duplicate it, please. You can put a `static inline` function in a header file if needed.

File: ntoskrnl/mm/balance.c [hunk 10]
```diff
@@ -39,6 +39,18 @@ static LONG PageOutThreadActive;
 
 /* FUNCTIONS ****************************************************************/
 
+static inline VOID UpdatePeakCommitment(VOID)
+{
+    // HACK: MmTotalCommittedPages should be adjusted consistently with
+    // other counters at different places.
+    MmTotalCommittedPages = MiMemoryConsumers[MC_SYSTEM].PagesUsed +
```
Comment (ThFabba) [inline ntoskrnl/mm/balance.c:46]: Does the caller of this function guarantee that some kind of global lock is held? If yes, please document that (or assert in code that the lock is held if possible, as well as use SAL to annotate the function).
If there's no lock, this should likely be an `InterlockedCompareExchange` loop. Depending on how critical the accuracy of this value is (is it used to enforce memory limits or only for returning information to user mode callers?), the loop might have to repeat until it observes that none of the three source values have changed during the calculation.

**Comment by ThFabba** [comment]
> Asking review to @ThFabba @tkreuzer @zefklop regarding the correctness of where the MmTotal...Pages counter needs to be modified in order to fairly represent the current total number of pages allocated.

I don't know about Windows comaptibility -- but given the three source values we currently calculate it from, updating it in the locations where those values change, as seems to be done here, makes sense to me.

File: ntoskrnl/mm/balance.c [hunk 11]
```diff
@@ -284,6 +296,7 @@ MmRequestPageMemoryConsumer(ULONG Consumer, BOOLEAN CanWait,
 
     /* Update the target */
     InterlockedIncrementUL(&MiMemoryConsumers[Consumer].PagesUsed);
+    UpdatePeakCommitment();
```
Comment (ThFabba) [inline ntoskrnl/mm/balance.c:299]: Actually, I think we're missing a matching update in `MmReleasePageMemoryConsumer` where this gets decremented again.

File: ntoskrnl/mm/pagefile.c [hunk 12]
```diff
@@ -330,7 +340,7 @@ MmAllocSwapPage(VOID)
             MiUsedSwapPages++;
             MiFreeSwapPages--;
             KeReleaseGuardedMutex(&MmPageFileCreationLock);
-
+            UpdatePeakCommitment();
```
Comment (ThFabba) [inline ntoskrnl/mm/pagefile.c:343]: `MmFreeSwapPage` should also call this, to keep the `Total` value accurate

File: ntoskrnl/mm/balance.c (hunk 10 above)
Comment (ThFabba) [reply ntoskrnl/mm/balance.c:46]: Actually, given how the `PagesUsed` and `MiUsedSwapPages` counters are updated, perhaps this function should just take a parameter to indicate increment/decrement, and do an `InterlockedIncrement` or `InterlockedDecrement` on the Total value. Much simpler than turning this into a loop or adding a lock.

File: ntoskrnl/mm/balance.c [hunk 13]
```diff
@@ -39,6 +39,18 @@ static LONG PageOutThreadActive;
...
+static inline VOID UpdatePeakCommitment(VOID)
+{
+    // HACK: MmTotalCommittedPages should be adjusted consistently with
+    // other counters at different places.
+    MmTotalCommittedPages = MiMemoryConsumers[MC_SYSTEM].PagesUsed +
+                          MiMemoryConsumers[MC_USER].PagesUsed +
+                          MiUsedSwapPages;
+
+    if (MmTotalCommittedPages > MmPeakCommitment)
+        MmPeakCommitment = MmTotalCommittedPages;
```
Comment (ThFabba) [inline ntoskrnl/mm/balance.c:51]: Related to the locking question above... atomically updating a maximum value looks something like this:
```
SIZE_T TotalCommittedPages = ...; // obtain this value atomically above instead of re-reading from memory
SIZE_T PeakCommitment = MmPeakCommitment;
while (TotalCommittedPages > PeakCommitment &&
       InterlockedCompareExchangeSizeT(&MmPeakCommitment, TotalCommittedPages, PeakCommitment) != PeakCommitment)
{
    PeakCommitment = MmPeakCommitment;
}
```

File: ntoskrnl/ex/sysinfo.c (hunk 6 above)
Comment (KRosUser) [reply ntoskrnl/ex/sysinfo.c:721]: Yes

File: ntoskrnl/mm/balance.c (hunk 7 above)
Comment (KRosUser) [reply ntoskrnl/mm/balance.c:42]: ok

File: ntoskrnl/mm/balance.c (hunk 11 above)
Comment (KRosUser) [reply ntoskrnl/mm/balance.c]: yes

File: ntoskrnl/mm/pagefile.c (hunk 12 above)
Comment (KRosUser) [reply ntoskrnl/mm/pagefile.c]: yes

File: ntoskrnl/mm/pagefile.c (hunk 9 above)
Comment (KRosUser) [reply ntoskrnl/mm/pagefile.c]: I used it once now with a ` static inline` declaration in a new mm.h but now it no longer compiles due to : 

```
In file included from ../ntoskrnl/mm/balance.c:18:
../ntoskrnl/mm/mm.h:2:20: error: 'UpdateTotalCommittedPages' used but never defined [-Werror]
 static inline VOID UpdateTotalCommittedPages(VOID);
                    ^~~~~~~~~~~~~~~~~~~~~~~~~
cc1.exe: all warnings being treated as errors
```

Could you help me please ?

Comment (katahiromz) [reply ntoskrnl/mm/pagefile.c]: inline function `UpdateTotalCommittedPages` needs its definition. Where is the contents?

File: ntoskrnl/mm/balance.c (hunk 10 above)
Comment (KRosUser) [reply ntoskrnl/mm/balance.c:50]: the "lock" discussion goes beyond my understanding of this module.
The current criticity is very low (display of "peak" data in TaskMgr), so i'd propose to go on with this "first simple" implementation and let another more experienced dev continue with another PR to make it safer and reader for more critical use case, if any. Would it make sense to you @ThFabba & @HBelusca ? Otherwise, i'd need more explicit contributions from you to this PR on this specific item :-)

File: ntoskrnl/mm/balance.c (hunk 13 above)
Comment (KRosUser) [reply ntoskrnl/mm/balance.c:55]: See previous comment / question regarding lock

File: ntoskrnl/mm/pagefile.c (hunk 9 above)
Comment (KRosUser) [reply ntoskrnl/mm/pagefile.c]: defined in mm.h and implemented in balance.c

Comment (KRosUser) [reply ntoskrnl/mm/pagefile.c]: (sorry i'm not familiar with `static inline`)

Comment (KRosUser) [reply ntoskrnl/mm/pagefile.c]: Done

File: ntoskrnl/mm/mm.h [hunk 14]
```diff
@@ -0,0 +1,2 @@
+#pragma once
+static inline VOID UpdateTotalCommittedPages(VOID);
```
Comment (HBelusca) [inline ntoskrnl/mm/mm.h:2]: Why creating this file, instead of using the existing `ntoskrnl/include/internal/mm.h` ?
Also, it's in that .h that the implementation of the inline UpdateTotalCommittedPages() function must be placed.

File: ntoskrnl/mm/pagefile.c [hunk 15]
```diff
@@ -31,6 +31,7 @@
 #include <ntoskrnl.h>
 #define NDEBUG
 #include <debug.h>
+#include "mm.h"
```
Comment (HBelusca) [inline ntoskrnl/mm/pagefile.c:34]:
```suggestion
```

File: ntoskrnl/mm/balance.c [hunk 16]
```diff
@@ -15,8 +15,7 @@
 #include <debug.h>
 
 #include "ARM3/miarm.h"
-
-
+#include "mm.h"
```
Comment (HBelusca) [inline ntoskrnl/mm/balance.c:18]: Please revert this change as well.

File: ntoskrnl/mm/balance.c [hunk 17]
```diff
@@ -39,6 +38,23 @@ static LONG PageOutThreadActive;
 
 /* FUNCTIONS ****************************************************************/
 
+static inline VOID UpdateTotalCommittedPages(VOID)
+{
+    /*
+     *   Add up all the used "Committed" memory + pagefile.
+     *   Not sure this is right. 8^\
+     *   MmTotalCommittedPages should be adjusted consistently with
+    *    other counters at different places.
```
Comment (HBelusca) [inline ntoskrnl/mm/balance.c:47]:
```suggestion
     * Add up all the used "Committed" memory + pagefile.
     * Not sure this is right. 8^\
     * MmTotalCommittedPages should be adjusted consistently with
     * other counters at different places.
```

File: ntoskrnl/mm/balance.c [hunk 18]
```diff
@@ -39,6 +38,23 @@ static LONG PageOutThreadActive;
 
 /* FUNCTIONS ****************************************************************/
 
+static inline VOID UpdateTotalCommittedPages(VOID)
```
Comment (HBelusca) [inline ntoskrnl/mm/balance.c:41]: That inline function should be moved into the mm.h (and the "static" be removed).

File: ntoskrnl/mm/mm.h (hunk 14 above)
Comment (KRosUser) [reply ntoskrnl/mm/mm.h:2]: ok

File: ntoskrnl/mm/pagefile.c (hunk 15 above)
Comment (KRosUser) [reply ntoskrnl/mm/pagefile.c:34]: ok

File: ntoskrnl/mm/balance.c (hunk 16 above)
Comment (KRosUser) [reply ntoskrnl/mm/balance.c:18]: ok

File: ntoskrnl/mm/balance.c (hunk 18 above)
Comment (KRosUser) [reply ntoskrnl/mm/balance.c:41]: ok

File: ntoskrnl/mm/pagefile.c [hunk 19]
```diff
@@ -31,6 +31,7 @@
 #include <ntoskrnl.h>
 #define NDEBUG
 #include <debug.h>
+#include "../include/internal/mm.h"
```
Comment (HBelusca) [inline ntoskrnl/mm/pagefile.c:34]: This header should already be automatically included, so this line should be not needed.

File: ntoskrnl/mm/balance.c [hunk 20]
```diff
@@ -15,8 +15,7 @@
 #include <debug.h>
 
 #include "ARM3/miarm.h"
-
-
+#include "../include/internal/mm.h"
```
Comment (HBelusca) [inline ntoskrnl/mm/balance.c:18]: Ditto.

Comment (KRosUser) [reply ntoskrnl/mm/balance.c]: Done

File: ntoskrnl/mm/pagefile.c (hunk 19 above)
Comment (KRosUser) [reply ntoskrnl/mm/pagefile.c]: Done

File: ntoskrnl/mm/balance.c (hunk 10 above)
Comment (ThFabba) [reply ntoskrnl/mm/balance.c]: You can just use my second suggestion, it's much simpler. Something like this should do the trick -- and then you call it with either +1 or -1 depending on whether the calling code just increased or decreased the commitment.
```
inline VOID UpdateTotalCommittedPages(LONG Increase)
{
    SIZE_T TotalCommittedPages = InterlockedExchangeAddSizeT(&MmTotalCommittedPages, Increase) + Increase;

    SIZE_T PeakCommitment = MmPeakCommitment;
    while (TotalCommittedPages > PeakCommitment &&
           InterlockedCompareExchangeSizeT(&MmPeakCommitment, TotalCommittedPages, PeakCommitment) != PeakCommitment)
    {
        PeakCommitment = MmPeakCommitment;
    }
}
```

File: ntoskrnl/mm/balance.c [hunk 21]
```diff
@@ -15,8 +15,6 @@
 #include <debug.h>
 
 #include "ARM3/miarm.h"
```
Comment (HBelusca) [inline ntoskrnl/mm/balance.c:17]: Restoring original newlines
```suggestion
#include "ARM3/miarm.h"


```

**Review by HBelusca — COMMENTED** [review]
Missing: the lock-free considerations from Thomas :thinking:

**Review by zefklop — APPROVED** [review]
I'd say it's good enough for now. There are other places where we should update those values, but that would be for later.

**Comment by KRosUser** [comment]
> I'd say it's good enough for now.

You mean no need for the lock stuff ? @HBelusca ?

**Comment by zefklop** [comment]
> 

Oh I didn't see this part of the conversation. Reading.

File: ntoskrnl/mm/balance.c (hunk 10 above)
Comment (zefklop) [reply ntoskrnl/mm/balance.c]: Yes, please use @ThFabba suggestion. At first I thought this was done under PFN lock, but that's obviously not the case.

**Comment by KRosUser** [comment]
> > 
> 
> Oh I didn't see this part of the conversation. Reading.

OK

File: ntoskrnl/mm/pagefile.c [hunk 22]
```diff
@@ -329,8 +330,9 @@ MmAllocSwapPage(VOID)
             }
             MiUsedSwapPages++;
             MiFreeSwapPages--;
-            KeReleaseGuardedMutex(&MmPageFileCreationLock);
+            UpdateTotalCommittedPages();
```
Comment (ThFabba) [inline ntoskrnl/mm/pagefile.c:333]: These two calls were good outside of the lock, like you had them before. That is, assuming we're making the updates interlocked.

File: ntoskrnl/mm/balance.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (SergeGautherie) [reply ntoskrnl/mm/balance.c]: My bad for leading you to remove this call, as we want to recalculate the total nonetheless, as Thomas pointed out.

File: ntoskrnl/mm/balance.c (hunk 10 above)
Comment (SergeGautherie) [reply ntoskrnl/mm/balance.c]: > LONG Increase

Good point. To further optimize, we could have 2 functions, as decrease never updates peak.

Comment (HBelusca) [reply ntoskrnl/mm/balance.c]: That's a pointless optimization. Just doing `if (Increase < 0) return;` before the while() loop should be sufficient.

Comment (SergeGautherie) [reply ntoskrnl/mm/balance.c]: My suggestion is to reduce compiled code size and increase execution speed. (And I find it more explicit / simpler.)
As always, do it as you prefer..

Comment (HBelusca) [reply ntoskrnl/mm/balance.c]: > to reduce compiled code size and increase execution speed

So we understand you actually did all those verifications before proposing your suggestions?

File: ntoskrnl/mm/balance.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (KRosUser) [reply ntoskrnl/mm/balance.c]: No problem, I put it back since then.

File: ntoskrnl/mm/balance.c (hunk 10 above)
Comment (KRosUser) [reply ntoskrnl/mm/balance.c]: Implementation done. I'm testing and i'll commit after if works fine.

**Comment by KRosUser** [comment]
"Locked" implementation as proposed by @ThFabba has been implemented with the optimization suggested by @HBelusca 
Tested and ready to go !

File: ntoskrnl/include/internal/mm.h [hunk 23]
```diff
@@ -866,6 +866,36 @@ NTAPI
 MmDeleteKernelStack(PVOID Stack,
                     BOOLEAN GuiStack);
 
+/* balance.c / pagefile.c******************************************************/
+
+inline VOID UpdateTotalCommittedPages(LONG Increase)
```
Comment (zefklop) [inline ntoskrnl/include/internal/mm.h:871]: Can you call that `Delta` or similar please ? Negative increase  is not an increase anymore 🙂

File: ntoskrnl/include/internal/mm.h [hunk 24]
```diff
@@ -866,6 +866,36 @@ NTAPI
...
+                            MiUsedSwapPages;
+     
+     */
+    
+    /* Update Commitment */
+    SIZE_T TotalCommittedPages = InterlockedExchangeAddSizeT(&MmTotalCommittedPages, Increase) + Increase;
+
+    SIZE_T PeakCommitment = MmPeakCommitment;
+    
+    if (Increase < 0) return; //if we decrement, we exit here to optimize processing (peak not impacted)
```
Comment (zefklop) [inline ntoskrnl/include/internal/mm.h:889]: If that's a nagative, yout guaranteed not to enter the while loop below.

Comment (KRosUser) [reply ntoskrnl/include/internal/mm.h]: Done

File: ntoskrnl/include/internal/mm.h (hunk 23 above)
Comment (KRosUser) [reply ntoskrnl/include/internal/mm.h]: Done

File: ntoskrnl/mm/pagefile.c (hunk 22 above)
Comment (KRosUser) [reply ntoskrnl/mm/pagefile.c]: Done

**Review by zefklop — APPROVED** [review]
Approving again. Looks better now.

File: ntoskrnl/include/internal/mm.h [hunk 25]
```diff
@@ -866,6 +866,34 @@ NTAPI
...
+    SIZE_T TotalCommittedPages = InterlockedExchangeAddSizeT(&MmTotalCommittedPages, Delta) + Delta;
+
+    SIZE_T PeakCommitment = MmPeakCommitment;
+
+    /* Update Peak */
+    while (TotalCommittedPages > PeakCommitment &&
+           InterlockedCompareExchangeSizeT(&MmPeakCommitment, TotalCommittedPages, PeakCommitment) != PeakCommitment)
+    {
+        PeakCommitment = MmPeakCommitment;
+    }
```
Comment (HBelusca) [inline ntoskrnl/include/internal/mm.h:894]: Side remark: an explanation of this pattern can be found e.g. at
https://devblogs.microsoft.com/oldnewthing/20040915-00/?p=37863

File: ntoskrnl/include/internal/mm.h [hunk 26]
```diff
@@ -866,6 +866,34 @@ NTAPI
...
+     * MmTotalCommittedPages should be adjusted consistently with
+     * other counters at different places.
+         MmTotalCommittedPages = MiMemoryConsumers[MC_SYSTEM].PagesUsed +
+                            MiMemoryConsumers[MC_USER].PagesUsed +
+                            MiUsedSwapPages;
+     
+     */
+    
+    /* Update Commitment */
+    SIZE_T TotalCommittedPages = InterlockedExchangeAddSizeT(&MmTotalCommittedPages, Delta) + Delta;
```
[the remark this answers is not in the document]
Comment (ThFabba) [reply ntoskrnl/include/internal/mm.h:885]: `InterlockedExchangeAdd` returns the "old" value. We want the new value, which is larger by `Delta`

File: ntoskrnl/include/internal/mm.h (hunk 23 above)
Comment (ThFabba) [reply ntoskrnl/include/internal/mm.h]: Thanks, I was struggling to come up with a better name haha

File: ntoskrnl/mm/pagefile.c [hunk 27]
```diff
@@ -329,8 +330,9 @@ MmAllocSwapPage(VOID)
             }
             MiUsedSwapPages++;
             MiFreeSwapPages--;
-            KeReleaseGuardedMutex(&MmPageFileCreationLock);
+            UpdateTotalCommittedPages(1);
```
Comment (ThFabba) [inline ntoskrnl/mm/pagefile.c:333]: As I think I said somewhere before, this can be called outside of the page file creation lock. But NBD either way.

**Review by ThFabba — APPROVED** [review]

File: ntoskrnl/include/internal/mm.h (hunk 26 above)
[the remark this answers is not in the document]
Comment (HBelusca) [reply ntoskrnl/include/internal/mm.h:885]: Aha good point :+1:

File: ntoskrnl/mm/pagefile.c (hunk 27 above)
Comment (KRosUser) [reply ntoskrnl/mm/pagefile.c:333]: Would you like me to swap both lines or is it ok for approval?

File: ntoskrnl/include/internal/mm.h (hunk 25 above)
Comment (KRosUser) [reply ntoskrnl/include/internal/mm.h:894]: Do you want me to add this as comment in code ?

File: ntoskrnl/mm/pagefile.c (hunk 27 above)
Comment (HBelusca) [reply ntoskrnl/mm/pagefile.c:333]: I'm personally fine with updating all counters at once (instead of having some counters with the lock held, then the other counter(s) with the lock unheld).

Comment (KRosUser) [reply ntoskrnl/mm/pagefile.c:333]: So I keep it as is since it's ok for @ThFabba too.

I let you answer on the comment and we're done !

File: ntoskrnl/include/internal/mm.h (hunk 25 above)
Comment (HBelusca) [reply ntoskrnl/include/internal/mm.h:893]: No, it was just for your (and possible other readers) personal knowledge.

Comment (KRosUser) [reply ntoskrnl/include/internal/mm.h:893]: Ok, thanks!

**Review by HBelusca — APPROVED** [review]

File: ntoskrnl/mm/balance.c (hunk 10 above)
Comment (SergeGautherie) [reply ntoskrnl/mm/balance.c]: > you actually did all those verifications

No, I simply assume that less parameters/variables/code in C results in less binary code.

Comment (ThFabba) [reply ntoskrnl/mm/balance.c]: Are you forgetting that you'd also be duplicating the function prologue/epilogue, alignment and the InterlockedExchangeAdd? It's not at all obvious that it would result in smaller code size - and it definitely wouldn't result in less C code, so I'm not sure how you get to this conclusion.
